src/sqlfluff/core/parser/segments/base.py

```diff
@@ -281,52 +275,79 @@ def expand(segments, parse_context):
-            for seg in todo_buffer:
-                if not seg.is_meta:
-                    running_pos = seg.pos_marker
+    @classmethod
+    def _position_segments(cls, segments, parent_pos=None):
+        """Refresh positions of segments within a span.
+
+        This does two things:
+        - Assign positions to any segments without them.
+        - Updates the working line_no and line_pos for all
+          segments during fixing.
+
+        New segments are assumed to be metas or insertions
+        and so therefore have a zero-length position in the
+        source and templated file.
+        """
+        # If there are no segments, there's no need to reposition.
+        if not segments:
+            return segments
+
+        # Work out our starting position for working through
+        if parent_pos:
+            line_no = parent_pos.working_line_no
+            line_pos = parent_pos.working_line_pos
+        # If we don't have it, infer it from the first position
+        # in this segment that does have a position.
+        else:
+            for fwd_seg in segments:
+                if fwd_seg.pos_marker:
+                    line_no = fwd_seg.pos_marker.working_line_no
+                    line_pos = fwd_seg.pos_marker.working_line_pos
                     break
             else:
-                raise ValueError("No starting pos provided and unable to infer.")
-
-        # Strip the starting position regardless. This means
-        # we don't accidentally contaminate any inserted initial
-        # segments.
-        running_pos = running_pos.strip()
-
-        while len(todo_buffer) > 0:
-            # Get the first off the buffer
-            seg = todo_buffer.pop(0)
-
-            # We'll preserve statement indexes so we should keep track of that.
-            # When recreating, we use the DELTA of the index so that's what matter...
-            idx = seg.pos_marker.statement_index - running_pos.statement_index
-            new_pos = seg.pos_marker.shift_to(running_pos)
-
-            if not meta_only or seg.is_meta:
-                # Copy the segment
-                seg_copy = copy.copy(seg)
-                # Update the position
-                seg_copy.pos_marker = new_pos
-                # Realign the children of that class if required.
-                if len(seg_copy.segments) > 0:
-                    seg_copy = seg_copy.realign()
-                seg = seg_copy
-
-            # Update the running position with the content of that segment.
-            running_pos = running_pos.advance_by(raw=seg.raw, idx=idx)
-            # Add the buffer to my new segment
-            seg_buffer.append(seg)
-        return tuple(seg_buffer)
+                linter_logger.warning("SEG: %r, POS: %r", segments, parent_pos)
+                raise ValueError("Unable to find working position.")
+
+        # Use the index so that we can look forward
+        # and backward.
+        for idx in range(len(segments)):
```

Comment:
Nitpick: Since the loop body frequently refers to `segments[idx]`, should we consider changing the loop to:

```
for idx, segment in enumerate(segments):
```

Then, all references to `segments[idx]` could be replaced with `segment`. This is. not a huge deal, but it seems more Pythonic and slightly more readable.

Reply:
Excellent idea!